### PR TITLE
IOSXE/show wlan id <id_number> client stats

### PR DESCRIPTION
**Pull request by robschn**
## Description
Adds parser for `show wlan id <id_number> client stats`.

## Motivation and Context
Takes raw output of `show wlan id <id_number> client stats` and coverts it to structured data.

## Impact (If any)
None

## Screenshots:
image

## Checklist:
- [X] I have updated the changelog.
- [X] I have updated the documentation (If applicable).
- [X] I have added tests to cover my changes (If applicable).
- [X] All new and existing tests passed.
- [X] All new code passed compilation.

### CODE REVIEW

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        "iapp_disassociate_wired": 0,
+        "idle_timeout": 0,
+        "internal_error": 0,
+        "ip_theft": 0,
+        "ip_update_timeout": 0,
+        "ip_zone_change": 0,
+        "l3_auth_fail": 0,
+        "mac": {"auth_fail": 0, "ip_theft": 0, "theft": 0},
+        "manually_excluded": 0,
+        "mobility": {"fail": 0, "peer_delete": 0, "tunnel_down": 0, "wlan_down": 0},
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py:160]: maybe `fail` need to be outside of `mobility`?
```
   Due to mobility failure                                         : 0
```
```
"due_to_mobility_failure": 0,
"mobility": {
    (snip)

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            "reset": 0,
+            "rule_error": 0,
+        },
+        "fqdn": {
+            "filter_missing": 0,
+            "invalid_group_id": 0,
+            "wrong_postauth_filter": 0,
+            "wrong_preauth_filter": 0,
+        },
+        "ha_recovery": 0,
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py:150]: `client_deleted_during_ha_recovery`

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            "slot_disable": 0,
+            "stop_reassociation": 0,
+            "transmission_deauth": 0,
+            "transmission_fail_threshold": 0,
+            "uplink_recieve_timeout": 0,
+            "vlan_delete": 0,
+        },
+        "avc_client_reanchor": 0,
+        "bad_credentials": 16,
+        "blacklist": 0,
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py:48]: `client_blacklist`

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            "invalid_parameter": 0,
+            "policy_fail": 0,
+            "send_ap_fail": 0,
+            "unbind_ap_fail": 0,
+        },
+        "radio_down": 0,
+        "reauth_fail": 0,
+        "reauthentication_fail": 0,
+        "redirect_acl_fail": 2,
+        "roam_fail": 0,
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py:187]: `inter_instance_roam_failure`

File: src/genie/libs/parser/iosxe/show_wlan.py [hunk 5]
```diff
@@ -104,3 +103,744 @@ def cli(self, output=None):
...
+                "qos",
+                "wired",
+            ]
+
+            for key in key_list:
+                new_key_group = {key: {}}
+
+                for item in new_group["client_delete"].copy():
+                    # if the key from key_list is found in item
+                    if re.search(f"^{key}_", item):
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/show_wlan.py:834]: sorry, we supported python3.5 until very recently. so we don't use f-string yet for users who have not migrated yet.

**Review by tahigash — CHANGES_REQUESTED** [review]
sorry, most of them are about key name. I understand you tried to use shorter name, but sometimes shorter name is, hard to guess or match to actual output when needed.

So, our approach is, we usually take the output as is as key name, but if it's very long, we consider changing it to shorter one.

File: src/genie/libs/parser/iosxe/show_wlan.py [hunk 6]
```diff
@@ -104,3 +104,744 @@ def cli(self, output=None):
...
+# ==============================
+# Parser for:
+#  * 'show wlan id client stats'
+# ==============================
+class ShowWlanIdClientStats(ShowWlanIdClientStatsSchema):
+    """Parser for show wlan id client stats"""
+
+    cli_command = "show wlan id {id_number} client stats"
+
+    def cli(self, id_number="", output=None):
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/show_wlan.py]: Updated! I changed around the if statements as well.

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        "vlan_change": 0,
+        "vlan_fail": 0,
+        "web_auth_fail": 0,
+        "wired": {"vlan_change": 0, "wbg_joins": 0, "wgb_change": 0, "wgb_delete": 0},
+        "wpa_groupkey_timeout": 9,
+        "wpa_key_timeout": 11,
+        "wrong_psk": 0,
+        "wrong_replay_counter": 6,
+    },
+    "client_stats": {"auth": 0, "ip_learn": 0, "mobility": 0, "run": 24, "webauth": 0},
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+        "session_timeout": 0,
+        "ssid_change": 0,
+        "supplicant_request": 153,
+        "supplicant_restart": 0,
+        "unknown": 0,
+        "vlan_change": 0,
+        "vlan_fail": 0,
+        "web_auth_fail": 0,
+        "wired": {"vlan_change": 0, "wbg_joins": 0, "wgb_change": 0, "wgb_delete": 0},
+        "wpa_groupkey_timeout": 9,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+            "invalid_akm": 0,
+            "invalid_ft_ie": 0,
+            "invalid_mdie": 0,
+            "invalid_pairwise_cipher": 24,
+            "invalid_pkmid": 0,
+            "invalid_qos_parameter": 0,
+            "invalid_rsnie_capabilities": 0,
+            "max_sta": 0,
+            "qos_policy": 0,
+            "sae_auth_fail": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
+expected_output = {
+    "client_delete": {
+        "aaa_not_ready": 0,
+        "aaa_unavailable": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            "domain_shut": 0,
+            "empty_domain": 0,
+            "flex_no_gateway": 0,
+            "generic_join_fail": 0,
+            "ha_reconcile": 0,
+            "invalid_domain": 0,
+            "invalid_gateway": 0,
+            "invalid_vlan": 0,
+            "mobility_handoff_error": 0,
+            "onboarding_error": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        "idle_timeout": 0,
+        "internal_error": 0,
+        "ip_theft": 0,
+        "ip_update_timeout": 0,
+        "ip_zone_change": 0,
+        "l3_auth_fail": 0,
+        "mac": {"auth_fail": 0, "ip_theft": 0, "theft": 0},
+        "manually_excluded": 0,
+        "mobility": {"fail": 0, "peer_delete": 0, "tunnel_down": 0, "wlan_down": 0},
+        "nack_ifid_exists": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        "anchor": {
+            "call_admission_controller": 0,
+            "create_fail": 0,
+            "invalid_mobility_bssid": 0,
+            "no_memory": 0,
+            "static_ip_fail": 0,
+        },
+        "ap_down": 136,
+        "ap_upgrade": 0,
+        "apinit": {
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+            "4way_fail": 0,
+            "acl_mismatch": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+            "rssi_low_threshold": 0,
+            "smart_roam_fail": 0,
+            "timer_timeout": 0,
+        },
+        "dot11w_max_sa": 0,
+        "dot11x_auth_timeout": 0,
+        "dot11x_credential_fail": 0,
+        "dot1x_no_config": 0,
+        "eap": {
+            "bad_keyframe": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+            "invalid_key_type": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+        "no_operation": 0,
+        "not_allowed_roaming": 0,
+        "policy": {
+            "bind_fail": 0,
+            "fail": 0,
+            "manager_error": 0,
+            "manager_mismatch": 0,
+        },
+        "port_admin_disabled": 0,
+        "ptk_error": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+            "sensor_station_timeout": 0,
+            "slot_disable": 0,
+            "stop_reassociation": 0,
+            "transmission_deauth": 0,
+            "transmission_fail_threshold": 0,
+            "uplink_recieve_timeout": 0,
+            "vlan_delete": 0,
+        },
+        "avc_client_reanchor": 0,
+        "bad_credentials": 3,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+        "nack_ifid_mismatch": 0,
+        "nas_error": 0,
+        "no_ip": 11,
+        "no_operation": 0,
+        "not_allowed_roaming": 0,
+        "policy": {
+            "bind_fail": 0,
+            "fail": 0,
+            "manager_error": 0,
+            "manager_mismatch": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_2_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+        "l3_auth_fail": 0,
+        "mac": {"auth_fail": 0, "ip_theft": 0, "theft": 0},
+        "manually_excluded": 0,
+        "mobility": {"fail": 0, "peer_delete": 0, "tunnel_down": 0, "wlan_down": 0},
+        "nack_ifid_exists": 0,
+        "nack_ifid_mismatch": 0,
+        "nas_error": 0,
+        "no_ip": 93,
+        "no_operation": 0,
+        "not_allowed_roaming": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py (hunk 1 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py (hunk 2 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py (hunk 3 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py (hunk 4 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Updated!

File: src/genie/libs/parser/iosxe/show_wlan.py (hunk 5 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/show_wlan.py]: Good catch! f-strings are just second nature to me 😁. Update!

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+            "transmission_deauth": 0,
+            "transmission_fail_threshold": 0,
+            "uplink_recieve_timeout": 0,
+            "vlan_delete": 0,
+        },
+        "avc_client_reanchor": 0,
+        "bad_credentials": 16,
+        "blacklist": 0,
+        "bssid_down": 1,
+        "cleanup_wgb_roam": 0,
```
[the remark this answers is not in the document]
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Updated!

**Comment by robschn** [comment]
@tahigash All the keys are updated with your feedback. I really appreciate having a second pair of eyes on the data, everything blends together after the 50th line!

File: src/genie/libs/parser/iosxe/show_wlan.py [hunk 22]
```diff
@@ -104,3 +103,748 @@ def cli(self, output=None):
...
+
+            for key in key_list:
+                new_key_group = {key: {}}
+
+                for item in new_group["client_delete"].copy():
+                    format_key = "^{key}_".format(key=key)
+                    # if the key from key_list is found in item
+                    if re.search(format_key, item):
+                        # replace the key and update with new_dict
+                        new_key = re.sub(f"^{key}_", "", item)
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/show_wlan.py:840]: sorry, I found another f-string. please change this as well.

**Review by tahigash — CHANGES_REQUESTED** [review]
found another f-string. but I believe this is last from my review.

File: src/genie/libs/parser/iosxe/show_wlan.py (hunk 22 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/show_wlan.py]: AH! No my bad! I just dropped in the `format_key` variable I made above

**Review by tahigash — APPROVED** [review]
thanks for addressing my comments. schema looks good now.

**Comment by danielgraziano** [comment]
Hi @robschn, typically we do 1 regex per line (or common lines) and we loop over the lines to check if the regex matches each line. 

Can you please explain the benefits of having huge regex and searching through the entire output?

**Comment by robschn** [comment]
> Hi @robschn, typically we do 1 regex per line (or common lines) and we loop over the lines to check if the regex matches each line. 
> 
> 
> 
> Can you please explain the benefits of having huge regex and searching through the entire output?

Hey @danielgraziano ! The main benefit is that it makes formatting the data easier since all the keys and values are in one big dictionary. It does have a side benefit of ensuring none of the lines are missed since that would cause the regex to not match.

**Comment by danielgraziano** [comment]
> > Hi @robschn, typically we do 1 regex per line (or common lines) and we loop over the lines to check if the regex matches each line.
> > Can you please explain the benefits of having huge regex and searching through the entire output?
> 
> Hey @danielgraziano ! The main benefit is that it makes formatting the data easier since all the keys and values are in one big dictionary. It does have a side benefit of ensuring none of the lines are missed since that would cause the regex to not match.
> 
> I typically loop over line by line. I really only do `MULTILINE` when the output is this large.

What about the possibility that the output might not contain some lines that are defined in the regex? The regex wouldn't match then for that case either. I think it would be possible to keep the parser simple as most of the lines can be captured by one (all encompassing) regex. 

For example: `re.compile(r'^(?P<key>[\S\s]+\S) +: +(?P<value>\d)$')` and then replace all spaces in the key with underscores and change it to lowercase.

And have extra regex for special cases. 

What are your thoughts @robschn?

**Comment by robschn** [comment]
Yeah very true, I'm working on the assumption that the information in the output will not change. Let me play around with your idea and get back to you!

**Comment by robschn** [comment]
@danielgraziano I made the changes you requested, please let me know what you think!

@tahigash the schema is more closely related to the actual output now, please review! I can try to organize them into smaller dictionaries if need be, but I thought you should see what it looks like as is.

File: src/genie/libs/parser/iosxe/show_wlan.py [hunk 23]
```diff
@@ -1344,3 +1343,310 @@ def cli(self, output=None):
...
+        
+        header_group = {}
+
+        wlan_info_obj = {}
+
+        for line in out.splitlines():
+            line = line.strip()
+
+            if wlan_info_capture.match(line):
+                match = wlan_info_capture.match(line)
```
Comment (danielgraziano) [inline src/genie/libs/parser/iosxe/show_wlan.py:1628]: I did a little testing for runtime and its about 4-5 times faster to do it this way:
``` python
m = wlan_info_capture.match(line)
if m:
    group = m.groupdict()
    (snip)
```
Time with two matches: 4.39999857917428e-05
Time with one match, saved to a variable: 8.999952115118504e-06

**Review by danielgraziano — COMMENTED** [review]
Once @tahigash reviews the schema and the suggested change is made, this is good to go!

File: src/genie/libs/parser/iosxe/show_wlan.py (hunk 23 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/show_wlan.py]: 4 secs is a lot! I appreciate the testing, I've updated with newest commit

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
+expected_output = {
+    "wlan_info": {"profile_name": "lizzard_Global", "id": 17},
+    "current_client_state_statistics": {
+        "authenticating": 7,
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py:4]: I would suggest to have `wlan_id` as parent key.
```suggestion
expected_output = {
    "wlan_id": {
        17: {
            "profile_name": lizzard_Global",
            "current_client_state_statistics":
                "authenticatiing": 7,
```
this schema will be easy when we need to merge 2 parsed_output across different wlan_id to one.

**Review by tahigash — CHANGES_REQUESTED** [review]
reviewed schema and 1 comment

File: src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py (hunk 24 above)
Comment (robschn) [reply src/genie/libs/parser/iosxe/tests/ShowWlanIdClientStats/cli/equal/golden_output_expected.py]: Makes sense, updated!

**Review by tahigash — APPROVED** [review]
schema looks good now

**Review by danielgraziano — APPROVED** [review]

**Pull request merged**